execution-engine/README.markdown

```diff
@@ -0,0 +1,581 @@
+
+**NOT SUPPORTED**
+
+#### `fd_close`
+```rust
+fn fd_close(fd: Fd) -> Result<(), ErrNo>;
+fn fd_close(fd: u32) -> ErrNo;
+```
+Close the file descriptor `fd`.
+
+#### `fd_datasync`
+```rust
+fn fd_datasync(fd: Fd) -> Result<(), ErrNo>;
+fn fd_datasync(fd: u32) -> ErrNo;
+```
+**NOT SUPPORTED**
+
+#### `fd_fdstat_get`
+```rust
+fn fd_fdstat_get(fd: Fd) -> Result<FdStat, ErrNo>;
+fn fd_fdstat_get(fd: u32, fd_stat: u32) -> ErrNo;
+```
+Write the status of the file descriptor `fd` to the memory at `fd_stat`.
+
+#### `fd_fdstat_set_flags`
+```rust
+fn fd_fdstat_set_flags(fd: Fd, flags: FdFlags) -> Result<(), ErrNo>;
+fn fd_fdstat_set_flags(fd: u32, flags: u16) -> ErrNo;
+```
+Set the flags of the file descriptor `fd` to `flags`.
+WASM passes a `u32` as the `flag` parameter.
+
+#### `fd_fdstat_set_rights`
+```rust
+fn fd_fdstat_set_rights(fd: Fd, base: Rights, inheriting: Rights) -> Result<(), ErrNo>;
+fn fd_fdstat_set_rights(fd: u32, base: u64, inheriting: u64) -> ErrNo;
+```
+Set the base rights and inheriting rights of the file descriptor `fd` to `base` and `inheriting` respectively.
+
+#### `fd_filestat_get`
+```rust
+fn fd_filestat_get(fd: Fd) -> Result<FileStat, ErrNo>;
+fn fd_filestat_get(fd: u32, file_stat: u32) -> ErrNo;
+```
+Write the status of the file opened by the file descriptor `fd` to the memory at `fd_stat`.
+
+#### `fd_filestat_set_size`
+```rust
+fn fd_filestat_set_size(fd: Fd, size: FileSize) -> Result<(), ErrNo>;
+fn fd_filestat_set_size(fd: u32, size: u64) -> ErrNo;
+```
+Set the file size of the file opened by the file descriptor `fd` to the new size `size`. 
+It will zero-fill, if the new size is larger than the existing size,
+or truncate, if the new size is smaller.
+
+#### `fd_filestat_set_times`
+```rust
+fn fd_filestat_set_times(fd: Fd, atime: Timestamp, mtime: Timestamp, flags: SetTimeFlags) -> Result<(), ErrNo>;
+fn fd_filestat_set_times(fd: u32, atime: u64, mtime: u64, flags: u16) -> ErrNo;
+```
+Set the time(s) of the file opened by the file descriptor `fd` based on the `flags`:
+* If `ATIME_NOW` is set, set the last access time to the current time.
+* If `ATIME` is set, set the last access time to `atime`.
+* If `MTIME_NOW` is set, set the last modification time to the current time.
+* If `MTIME` is set, set the last modification time to `mtime`.
+
+Veracruz immediately returns `NoSys` if `ATIME_NOW` or `MTIME_NOW` is set.
+Veracruz internally does _NOT_ update the time information and also never use them.
+WASM passes a `u32` as the `flags` parameter.
+
+**QUESTION:** Priority of the flags.
+
+#### `fd_prestat_get`
+```rust
+fn fd_prestat_get(fd: Fd) -> Result<Prestat, ErrNo>;
+fn fd_prestat_get(fd: u32, pre_stat: u32) -> ErrNo;
+```
+Write the status of the pre-opened file descriptor `fd` to the memory at `pre_stat`. If the file descriptor is not pre-opened, return `BadF`.
+
+#### `fd_prestat_dir_name`
+```rust
+fn fd_prestat_dir_name(fd: Fd, path: Pointer<u8>, len: size) -> Result<(), ErrNo>;
```

Comment:
Also should `Pointer<u8>` be `AsRef<Path>`? Looks like the documentation got a bit out of sync from the code.

Reply:
I believe u8 might be a better representation here. Wasi uses bytes to represent any string and path. I actually copy-paste the first line from https://github.com/WebAssembly/WASI/blob/main/phases/snapshot/docs.md.

Reply:
Ah I see, that makes sense. It is a bit confusing now with the slightly different types, but I guess that not really a big deal.

Reply:
It is a goal for this document. WASI document and the actual API/ABI definition are different. I want to write it down in case people are reading the document [here](ttps://github.com/WebAssembly/WASI/blob/main/phases/snapshot/docs.md)  and work on the actual implementation of WASI.

Reply:
It is quite a useful document, and a good place to capture the quirks of WASI. Maybe the best documentation of the WASI functions you've covered that's out there? :)